src/bgw/job.c

```diff
@@ -567,6 +567,148 @@ ts_bgw_job_delete_by_id(int32 job_id)
+		repl[AttrNumberGetAttrOffset(Anum_bgw_job_schedule_interval)] = true;
+	}
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_max_runtime)] =
+		IntervalPGetDatum(&updated_job->fd.max_runtime);
+	repl[AttrNumberGetAttrOffset(Anum_bgw_job_max_runtime)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_max_retries)] =
+		Int32GetDatum(updated_job->fd.max_retries);
+	repl[AttrNumberGetAttrOffset(Anum_bgw_job_max_retries)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_retry_period)] =
+		IntervalPGetDatum(&updated_job->fd.retry_period);
+	repl[AttrNumberGetAttrOffset(Anum_bgw_job_retry_period)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_scheduled)] =
+		BoolGetDatum(updated_job->fd.scheduled);
+	repl[AttrNumberGetAttrOffset(Anum_bgw_job_scheduled)] = true;
+
+	repl[AttrNumberGetAttrOffset(Anum_bgw_job_config)] = true;
+	if (updated_job->fd.config)
+	{
+		ts_cm_functions->job_config_check(&updated_job->fd.proc_schema,
+										  &updated_job->fd.proc_name,
+										  updated_job->fd.config);
+		values[AttrNumberGetAttrOffset(Anum_bgw_job_config)] =
+			JsonbPGetDatum(updated_job->fd.config);
+	}
+	else
+		isnull[AttrNumberGetAttrOffset(Anum_bgw_job_config)] = true;
+
+	new_tuple = heap_modify_tuple(tuple, ts_scanner_get_tupledesc(ti), values, isnull, repl);
+
+	ts_catalog_update(ti->scanrel, new_tuple);
+
+	heap_freetuple(new_tuple);
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+/*
+ * Overwrite job with specified job_id with the given fields
+ *
+ * This function only updates the fields modifiable with alter_job.
+ */
+bool
+ts_bgw_job_update_by_id(int32 job_id, BgwJob *job)
+{
+	ScanKeyData scankey[1];
+	Catalog *catalog = ts_catalog_get();
+	ScanTupLock scantuplock = {
+		.waitpolicy = LockWaitBlock,
+		.lockmode = LockTupleExclusive,
+	};
+	ScannerCtx scanctx = { .table = catalog_get_table_id(catalog, BGW_JOB),
+						   .index = catalog_get_index(catalog, BGW_JOB, BGW_JOB_PKEY_IDX),
+						   .nkeys = 1,
+						   .scankey = scankey,
+						   .data = job,
+						   .limit = 1,
+						   .tuple_found = bgw_job_tuple_update_by_id,
+						   .lockmode = RowExclusiveLock,
+						   .scandirection = ForwardScanDirection,
+						   .result_mctx = CurrentMemoryContext,
+						   .tuplock = &scantuplock };
+
+	ScanKeyInit(&scankey[0],
+				Anum_bgw_job_pkey_idx_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(job_id));
+
+	return ts_scanner_scan(&scanctx);
+}
+
+static void
+ts_bgw_job_check_max_retries(BgwJob *job)
+{
+	BgwJobStat *job_stat;
+
+	job_stat = ts_bgw_job_stat_find(job->fd.id);
+
+	/* stop to execute failing jobs after reached the "max_retries" option */
+	if (job->fd.max_retries > 0 && job_stat->fd.consecutive_failures >= job->fd.max_retries)
+	{
+		ereport(WARNING,
+				(errcode(ERRCODE_WARNING),
+				 errmsg("job %d reached max_retries after %d consecutive failures",
+						job->fd.id,
+						job_stat->fd.consecutive_failures),
```

Comment:
I still think this is too much details for the error message.

Here is an example from the PostgreSQL source (`fd.c`):
```
				(errcode(ERRCODE_INSUFFICIENT_RESOURCES),
				 errmsg("insufficient file descriptors available to start server process"),
				 errdetail("System allows %d, we need at least %d.",
						   max_safe_fds + NUM_RESERVED_FDS,
						   FD_MINFREE + NUM_RESERVED_FDS)));
```

```suggestion
				 errmsg("job %d reached max_retries", job->fd.id),
```

Reply:
I changed it to include more details per @svenklemm suggestion

Reply:
I think the comment was more about grammar than that any specific information that should be there. Make a call on what you think is reasonable and go with it.